utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/AvroUtils.scala

```diff
@@ -105,31 +110,81 @@ object AvroUtils extends LazyLogging {
-      case (str: String, Schema.Type.ENUM) =>
-        new GenericData.EnumSymbol(schema, str)
+  def createAvroData(value: Any, schema: Schema): Any =
+    (value, schema.getType) match {
+      case (map: collection.Map[String@unchecked, _], Type.RECORD) =>
+        createRecord(schema, map)
+      case (map: util.Map[String@unchecked, _], Type.RECORD) =>
+        createRecord(schema, map.asScala)
+      case (collection: Traversable[_], Type.ARRAY) =>
+        collection.map(createAvroData(_, schema.getElementType)).toList.asJava
+      case (collection: util.Collection[_], Type.ARRAY) =>
+        createAvroData(collection.asScala, schema)
+      case (map: collection.Map[String@unchecked, _], Type.MAP) =>
+        map.mapValues(createAvroData(_, schema.getValueType)).asJava
+      case (map: util.Map[String@unchecked, _], Type.MAP) =>
+        createAvroData(map.asScala, schema)
+      case (str: String, Type.ENUM) =>
+        new EnumSymbol(schema, str)
+      case (str: String,Type.FIXED) =>
+        new Fixed(schema, str.getBytes("UTF-8"))
+      case (null, Type.UNION) if schema.isNullable => null
+      case (None, Type.UNION) if schema.isNullable => null
+      case (bytes: Array[Byte], Type.BYTES) =>
+        ByteBuffer.wrap(bytes)
+      case (any, Type.UNION) =>
+        schema.getTypes.asScala.filterNot(_.isNullable).map(createAvroData(any, _)).headOption.orNull
       case (_, _) => value
     }
 
+  /**
+    * It's a simply mapper scala Map[String, Any] to Avro GenericRecord
+    */
+  def createRecord(schema: Schema, data: scala.collection.Map[String, Any]): GenericRecord = {
+    val fields = schema.getFields.asScala.map(_.name()).toSet
     val builder = new LogicalTypesGenericRecordBuilder(schema)
-    data.foreach{ case (key, value) =>
 
-      val valueToSet = Option(schema.getField(key))
-        .map(field => createValue(value, field.schema()))
-        .getOrElse(throw new IllegalArgumentException(s"Unknown field $key in schema $schema."))
+    data
+      .filter{ case (key, _) => fields.contains(key) }
+      .foreach{ case (key, value) =>
+        val field = schema.getField(key)
+        val valueToSet = createAvroData(value, field.schema())
+
+        builder.set(key, valueToSet)
+      }
+
+    val results = builder.build()
+    results
+  }
+
+  def verifyLogicalType[T: ClassTag](schema: Schema): Boolean = {
+    val clazz = classTag[T].runtimeClass.asInstanceOf[Class[T]]
+    schema.getLogicalType != null && clazz.isAssignableFrom(schema.getLogicalType.getClass)
+  }
+
+  /**
+    * Discovering AvoSchema based on data
+    */
+  def getSchema(data: Any): Schema = {
+    def discoverSchema(data: List[Any]) = data.map(getSchema).distinct match {
+      case head :: Nil => head
+      case list => Schema.createUnion(list.asJava)
+    }
 
-      builder.set(key, valueToSet)
+    data match {
+      case container: GenericContainer =>
+        container.getSchema
+      case map: java.util.Map[_, _] =>
```

Comment:
Why you need to magically handle list of elements with schemas etc? It doesn't look like something straightforward

Reply:
Our `ConfluentUtils.serializeDataToBytesArray` accepts `schemaId` but inside method writer needs schema. There is no problem when we try to serialize record, because it contains own schema. Problem is when we try to serialize e.g. map of list - in that case we have to discover how schema looks.

Reply:
Still don't get it. Where you find this convention of determining schema based on list, map? I saw sth like `useSchemaReflection` in confluent's code, but have no idea  if we are still in generic records mode or sth else.

Reply:
Maybe you should use `NonRecordContainer` and pass this schema on your own?